File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/AddChirpTest.java
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.model.network.JsonTestUtils;
import com.github.dedis.popstellar.model.network.method.message.data.Action;
import com.github.dedis.popstellar.model.network.method.message.data.Objects;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.google.gson.JsonParseException;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageID;
import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageIDOtherThan;
import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
public class AddChirpTest {

  private static final String TEXT = "Hello guys";
  private static final MessageID PARENT_ID = generateMessageID();
  private static final long TIMESTAMP = 1631280815;

  private static final AddChirp ADD_CHIRP = new AddChirp(TEXT, PARENT_ID, TIMESTAMP);

  @Before
  public void setup() {
    JsonTestUtils.loadGSON(ApplicationProvider.getApplicationContext());
  }

  @Test
  public void createAddChirpWithTextTooLongTest() {
    String textTooLong =
        "This text should be way over three hundred characters which is the current limit of the"
            + " text within a chirp, and if I try to set the chirp's text to this, it  should"
            + " throw and IllegalArgumentException() so I hope it does otherwise I might have"
            + " screwed something up. But normally it is not that hard to write enough to reach"
            + " the threshold.";
    assertThrows(
        IllegalArgumentException.class, () -> new AddChirp(textTooLong, PARENT_ID, TIMESTAMP));
  }

  @Test
  public void getObjectTest() {
    assertEquals(Objects.CHIRP.getObject(), ADD_CHIRP.getObject());
  }

  @Test
  public void getActionTest() {
    assertEquals(Action.ADD.getAction(), ADD_CHIRP.getAction());
  }

  @Test
  public void getTextTest() {
    assertEquals(TEXT, ADD_CHIRP.getText());
  }

  @Test
  public void getParentIdTest() {
    assertEquals(PARENT_ID, ADD_CHIRP.getParentId().get());
  }

  @Test
  public void getTimestampTest() {
    assertEquals(TIMESTAMP, ADD_CHIRP.getTimestamp());
  }

  @Test
  public void equalsTest() {
    assertEquals(ADD_CHIRP, new AddChirp(TEXT, PARENT_ID, TIMESTAMP));

    String random = "random";
    assertNotEquals(ADD_CHIRP, new AddChirp(random, PARENT_ID, TIMESTAMP));
    assertNotEquals(
        ADD_CHIRP, new AddChirp(TEXT, generateMessageIDOtherThan(PARENT_ID), TIMESTAMP));
    assertNotEquals(ADD_CHIRP, new AddChirp(TEXT, PARENT_ID, TIMESTAMP + 1));
  }

  @Test
  public void jsonValidationTest() {
    JsonTestUtils.testData(ADD_CHIRP);

    String path =
        "protocol/examples/messageData/chirp_add_publish/wrong_chirp_add_publish_negative_time.json";
    String invalidJson = JsonTestUtils.loadFile(path);
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(invalidJson));
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/AddChirpTest.kt
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.model.network.JsonTestUtils.loadFile
import com.github.dedis.popstellar.model.network.JsonTestUtils.parse
import com.github.dedis.popstellar.model.network.JsonTestUtils.testData
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.google.gson.JsonParseException
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class AddChirpTest {
  @Test
  fun createAddChirpWithTextTooLongTest() {
    val textTooLong =
      ("This text should be way over three hundred characters which is the current limit of the" +
        " text within a chirp, and if I try to set the chirp's text to this, it  should" +
        " throw and IllegalArgumentException() so I hope it does otherwise I might have" +
        " screwed something up. But normally it is not that hard to write enough to reach" +
        " the threshold.")
    Assert.assertThrows(IllegalArgumentException::class.java) {
      AddChirp(textTooLong, PARENT_ID, TIMESTAMP)
    }
  }

  @Test
  fun objectTest() {
    Assert.assertEquals(Objects.CHIRP.`object`, ADD_CHIRP.`object`)
  }

  @Test
  fun actionTest() {
    Assert.assertEquals(Action.ADD.action, ADD_CHIRP.action)
  }

  @Test
  fun textTest() {
    Assert.assertEquals(TEXT, ADD_CHIRP.text)
  }

  @Test
  fun parentIdTest() {
    Assert.assertEquals(PARENT_ID, ADD_CHIRP.getParentId().get())
  }

  @Test
  fun timestampTest() {
    Assert.assertEquals(TIMESTAMP, ADD_CHIRP.timestamp)
  }

  @Test
  fun equalsTest() {
    Assert.assertEquals(ADD_CHIRP, AddChirp(TEXT, PARENT_ID, TIMESTAMP))
    val random = "random"
    Assert.assertNotEquals(ADD_CHIRP, AddChirp(random, PARENT_ID, TIMESTAMP))
    Assert.assertNotEquals(
      ADD_CHIRP,
      AddChirp(TEXT, Base64DataUtils.generateMessageIDOtherThan(PARENT_ID), TIMESTAMP)
    )
    Assert.assertNotEquals(ADD_CHIRP, AddChirp(TEXT, PARENT_ID, TIMESTAMP + 1))
  }

  @Test
  fun jsonValidationTest() {
    testData(ADD_CHIRP)
    val path =
      "protocol/examples/messageData/chirp_add_publish/wrong_chirp_add_publish_negative_time.json"
    val invalidJson = loadFile(path)
    Assert.assertThrows(JsonParseException::class.java) { parse(invalidJson) }
  }

  companion object {
    private const val TEXT = "Hello guys"
    private val PARENT_ID = Base64DataUtils.generateMessageID()
    private const val TIMESTAMP: Long = 1631280815
    private val ADD_CHIRP = AddChirp(TEXT, PARENT_ID, TIMESTAMP)
  }
}
